src/domain/mongoose_lazy_routing.erl

```diff
@@ -0,0 +1,397 @@
+                                              Component :: module()) ->
+    {ok, mongoose_iq_handler:t()} | {error, not_found}.
+handle_unregister_iq_handler_for_domain(HostType, Namespace, Component) ->
+    IQKey = #iq_table_key{host_type = HostType, namespace = Namespace,
+                          component = Component},
+    case ets:lookup(?IQ_TABLE, IQKey) of
+        [] -> {error, not_found};
+        [{_, IQHandler} = IQ] ->
+            Domains = ets:match_object(?ROUTING_TABLE, {'_', HostType}),
+            unregister_iqs([IQ], Domains),
+            {ok, IQHandler}
+    end.
+
+-spec handle_unregister_iq_handler_for_subdomain(HostType :: mongooseim:host_type(),
+                                                 SubdomainPattern :: subdomain_pattern(),
+                                                 Namespace :: binary(),
+                                                 Component :: module()) ->
+    {ok, mongoose_iq_handler:t()} | {error, not_found}.
+handle_unregister_iq_handler_for_subdomain(HostType, SubdomainPattern,
+                                           Namespace, Component) ->
+    IQKey = #iq_table_key{host_type = HostType, subdomain_pattern = SubdomainPattern,
+                          namespace = Namespace, component = Component},
+    case ets:lookup(?IQ_TABLE, IQKey) of
+        [] -> {error, not_found};
+        [{_, IQHandler} = IQ] ->
+            Domains = ets:match_object(?ROUTING_TABLE, {'_', HostType}),
+            unregister_iqs([IQ], Domains),
+            {ok, IQHandler}
+    end.
+
+-spec handle_maybe_add_domain_or_subdomain(domain()) -> boolean().
+handle_maybe_add_domain_or_subdomain(Domain) ->
+    case ets:member(?ROUTING_TABLE, Domain) of
+        true ->
+            %% It's absolutely normal situation. We can receive
+            %% a couple of maybe_add_domain_or_subdomain requests
+            %% from different processes, so this domain can be
+            %% already added.
+            true;
+        false ->
+            try_to_add_domain_or_subdomain(Domain)
+    end.
+
+-spec try_to_add_domain_or_subdomain(domain()) -> boolean().
+try_to_add_domain_or_subdomain(Domain) ->
+    case mongoose_domain_api:get_domain_host_type(Domain) of
+        {ok, HostType} ->
+            add_domain(HostType, Domain),
+            true;
+        {error, not_found} ->
+            case mongoose_domain_api:get_subdomain_info(Domain) of
+                {ok, Info} ->
+                    add_subdomain(Info),
+                    true;
+                {error, not_found} ->
+                    false
+            end
+    end.
+
+-spec add_domain(host_type(), domain()) -> ok.
+add_domain(HostType, Domain) ->
+    DomainElement = {Domain, HostType},
+    case ets:insert_new(?ROUTING_TABLE, DomainElement) of
+        true ->
+            IQs = get_iqs(#iq_table_key{host_type = HostType}),
+            register_iqs(IQs, [DomainElement]),
+            ejabberd_local:register_host(Domain),
```

Comment:
This is going to add hook handlers per domain, not host type, see the code in `ejabberd_local`

Reply:
ejabberd_local is about routing. It works with domains, not host types. the way how ejabberd_local triggers hooks must be reworked (and we can supply host_types here if required), but that is a topic for another PR.